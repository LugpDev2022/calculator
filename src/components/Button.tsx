import { Dispatch } from 'react';
import evaluateOperation from '../helpers/evaluateOperation';
import { CalcAction } from '../calculatorReducer';

interface Props {
  dispatch: Dispatch<CalcAction>;
  operation: string;
  keySign: string;
}

const Button: React.FC<Props> = ({ dispatch, operation, keySign }) => {
  const handleClick = () => {
    if (keySign === 'DEL') return dispatch({ type: 'DELETE' });

    if (keySign === 'AC') return dispatch({ type: 'RESET' });

    if (keySign === '=') {
      const { result, error } = evaluateOperation(operation);

      if (error) return dispatch({ type: 'SET_ERROR', payload: error });

      dispatch({ type: 'SET_RESULT', payload: result });

      return;
    }

    if (operation.length > 18) return;

    dispatch({ type: 'SET_OPERATION', payload: keySign });
  };

  return (
    <button
      onClick={handleClick}
      className='
        calculator-btn
        py-2
        rounded-lg
        sm:py-4
        font-bold
        text-3xl
      '
    >
      {keySign}
    </button>
  );
};

export default Button;
