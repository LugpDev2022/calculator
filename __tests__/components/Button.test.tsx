import React from 'react';
import { vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import Button from '../../src/components/Button';

describe('Tests on <Button />', () => {
  beforeEach(() => vi.clearAllMocks());

  test('should match snapshot', () => {
    const { container } = render(
      <Button dispatch={() => {}} keySign='DEL' operation='2+2' />
    );

    expect(container).toMatchSnapshot();
  });

  test('should call dispatch with delete', () => {
    const dispatchMock = vi.fn();

    render(<Button dispatch={dispatchMock} keySign='DEL' operation='2+2' />);

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(dispatchMock).toBeCalledWith({ type: 'DELETE' });
  });

  test('should call dispatch with reset', () => {
    const dispatchMock = vi.fn();

    render(<Button dispatch={dispatchMock} keySign='AC' operation='2+2' />);

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(dispatchMock).toBeCalledWith({ type: 'RESET' });
  });

  test('should call dispatch with set error on invalid operation', () => {
    const dispatchMock = vi.fn();

    render(<Button dispatch={dispatchMock} keySign='=' operation='10+7+5(' />);

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(dispatchMock).toBeCalledWith({
      type: 'SET_ERROR',
      payload: 'Invalid operation',
    });
  });

  test('should call dispatch with set result', () => {
    const dispatchMock = vi.fn();

    render(<Button dispatch={dispatchMock} keySign='=' operation='10+7+5' />);

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(dispatchMock).toBeCalledWith({ type: 'SET_RESULT', payload: 22 });
  });

  test('should call dispatch set operation', () => {
    const dispatchMock = vi.fn();

    render(
      <Button
        dispatch={dispatchMock}
        keySign='2'
        operation='10+7+52+523452+2342432'
      />
    );

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(dispatchMock).not.toHaveBeenCalled();
  });

  test('should call dispatch set operation', () => {
    const keySign = '2';
    const dispatchMock = vi.fn();

    render(
      <Button dispatch={dispatchMock} keySign={keySign} operation='10+7+5' />
    );

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(dispatchMock).toBeCalledWith({
      type: 'SET_OPERATION',
      payload: keySign,
    });
  });
});
