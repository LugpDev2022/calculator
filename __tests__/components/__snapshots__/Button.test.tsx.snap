// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests on <Button /> > should match snapshot 1`] = `
<div>
  <button
    class=" calculator-btn py-2 rounded-lg sm:py-4 font-bold text-3xl "
  >
    DEL
  </button>
</div>
`;
